runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/control/DefaultJobBundleFactory.java

```diff
@@ -352,20 +407,18 @@ public RemoteBundle getBundle(
         // The blocking queue of caches for serving multiple bundles concurrently.
         currentCache = availableCaches.take();
         client = currentCache.getUnchecked(executableStage.getEnvironment());
-        client.ref();
```

Comment:
`client.ref` needs to remain here, the lines below rely on that and it is also more readable.

Reply:
Please explain which lines rely on it.

Readability is highly highly subjective, I find it more readable to not duplicate the same code for two branches, as it is currently the case (which is potentially error-prone).

Reply:
I checked again and couldn't see how this change alters the behavior. In any case, I don't mind to move it back to unblock this PR.

Reply:
Below, `preparedClients.keySet().removeIf(c -> c.bundleRefCount.get() <= 0);` removes everything that isn't referenced. These two statements are logically one unit and hence I prefer to not scatter them:
```
client = currentCache.getUnchecked(executableStage.getEnvironment());
        client.ref();
client.ref();
```

Reply:
It's not true that a later `ref()` introduces a bug for `preparedClients.keySet().removeIf(c -> c.bundleRefCount.get() <= 0);` because the refcount will be >0, otherwise we wouldn't be able to retrieve the client from the cache.

In any case, I will revert this change.

Reply:
> It's not true that a later `ref()` introduces a bug for `preparedClients.keySet().removeIf(c -> c.bundleRefCount.get() <= 0);` because the refcount will be >0, otherwise we wouldn't be able to retrieve the client from the cache.

Cache and environment are shared between executable stages. So the refcount can become 0 with concurrent eviction and release. That actually raises the question if these 2 statements should be atomic.

Reply:
Yes, that makes sense. I've already reverted the change.

I suppose there is a race condition where we retrieve an environment X and before we can call `ref()` on it, we evict the environment X, close all its references, and shut it down. This will result in a job restart.